### PR TITLE
Cached DisplayList opacity inheritance fix

### PR DESCRIPTION
**Pull request by flar**
Fixes https://github.com/flutter/flutter/issues/120455

There is a small chance that the new FML_DCHECK will trigger in examples that we haven't seen yet, but that would point out more problems like this one that need to be fixed anyway.

This would be a good candidate to cherry-pick to any release that shipped with the original code in https://github.com/flutter/engine/pull/34562.

### CODE REVIEW

File: flow/raster_cache.cc [hunk 1]
```diff
@@ -112,14 +112,18 @@ bool RasterCache::UpdateCacheEntry(
 
 int RasterCache::MarkSeen(const RasterCacheKeyID& id,
```
Comment (chinmaygarde) [inline flow/raster_cache.cc:113]: Super minor nit: Not sure about the out parameter. Perhaps a tuple or struct return. But your call.

**Review by chinmaygarde — APPROVED** [review]

**Comment by chinmaygarde** [comment]
I agree this should be cherry-picked. Seems low risk and high impact.

File: flow/raster_cache.cc (hunk 1 above)
Comment (dnfield) [reply flow/raster_cache.cc:113]: Make this return `std::tuple<int, bool>`.

Callers can then do

```cpp
int accesses;
bool has_image;
std::tie(accesses, has_image) = foo->MarkSeen(...);
```

and you don't have the out parameter anymore.

**Comment by dnfield** [comment]
@zanderso do you know what the. malioc failure means?

**Comment by flar** [comment]
Rebased in an attempt to fix the malioc errors...

File: flow/raster_cache.cc (hunk 1 above)
Comment (flar) [reply flow/raster_cache.cc]: I don't like returning unnamed parameters like that. Having the caller name them is bad practice - they API should describe its own return values. I used a struct instead.

**Comment by flar** [comment]
Note that this fix creates a chicken-and-egg situation where an opacity parent with picture children that can only inherit opacity when they successfully cache themselves stalemate against each other in the decision as to who should do the caching. See https://github.com/flutter/flutter/issues/121740

**Pull request merged**